File: packages/ts/src/accounts/unified.ts
/**
 * Unified eth-native QoreChain accounts.
 *
 * ONE `eth_secp256k1` key = ONE 20-byte identity, rendered THREE ways so a wallet
 * never "has funds on one lane but not another". The 20 bytes are the Ethereum
 * derivation `keccak256(uncompressedPubkey[1:])[12:]`, so the key is natively
 * spendable on the EVM lane; the native (`qor1…`) and SVM (base58) forms are just
 * other encodings of the SAME 20 bytes — the chain reads one x/bank balance for
 * the account, visible under all three encodings.
 *
 * This is ADDITIVE. The legacy coin-118 native derivation
 * ({@link ../accounts/wallet.deriveNativeAccount}) is unchanged and still
 * supported; unified accounts are a distinct, opt-in identity model built on the
 * eth-native (coin-60) key.
 *
 * ```
 *   {
 *     mnemonic, privateKey (0x hex, 32B),
 *     publicKey (33B compressed, 0x hex),
 *     addressBytes (20B),
 *     cosmos: "qor1…",              // bech32
 *     evm:    "0x…" (EIP-55),       // hex
 *     svm:    "<base58>",           // base58(20B ‖ 12 zero bytes) = 32-byte addr
 *     pqc:    { publicKey, secretKey }  // ML-DSA-87 (Dilithium-5), for the hybrid ante
 *   }
 * ```
 *
 * Derivation uses audited primitives: @scure/bip39 (mnemonic + seed), @scure/bip32
 * (secp256k1 HD), @noble/curves (secp256k1), @noble/hashes (keccak256), @scure/base
 * (base58), and `@qorechain/pqc` (ML-DSA-87 + SHAKE-256). Secret material is
 * returned explicitly and never logged.
 */

import { mnemonicToSeed } from "@scure/bip39";
import { HDKey } from "@scure/bip32";
import { secp256k1 } from "@noble/curves/secp256k1";
import { keccak_256 } from "@noble/hashes/sha3";
import { base58 } from "@scure/base";
import { shake256 } from "@qorechain/pqc";

import { bytesToBech32, bech32ToHex } from "../utils/address";
import { validateMnemonic } from "./wallet";
import { generatePqcKeypair, type PqcKeypair } from "./pqc";

/** Bech32 human-readable prefix for QoreChain account addresses. */
const HRP = "qor";

/** Ethereum HD path (coin-type 60): makes the 20-byte address the keccak derivation. */
const ETH_HD_PATH_PREFIX = "m/44'/60'/0'/0/";

/**
 * The three address encodings of a single 20-byte QoreChain account, plus the
 * raw address bytes.
 */
export interface UnifiedAddresses {
  /** The raw 20-byte account address. */
  addressBytes: Uint8Array;
  /** Bech32 (`qor1…`) native encoding. */
  cosmos: string;
  /** `0x`-prefixed, EIP-55 mixed-case checksummed hex encoding. */
  evm: string;
  /** base58 of the 32-byte SVM address (`addr20 ‖ 12 zero bytes`). */
  svm: string;
}

/**
 * A fully derived unified eth-native account: one secp256k1 key rendered as all
 * three QoreChain address encodings, plus its deterministic ML-DSA-87 PQC keypair
 * for the hybrid ante.
 *
 * Secret fields (`mnemonic`, `privateKey`, `pqc.secretKey`) are returned so the
 * caller can sign; they are never logged by the SDK.
 */
export interface UnifiedAccount extends UnifiedAddresses {
  /**
   * The BIP-39 mnemonic this account was derived from, when derived from a
   * mnemonic. `undefined` for seed-derived accounts
   * ({@link unifiedAccountFromSeed}).
   */
  mnemonic?: string;
  /** `0x`-prefixed 32-byte secp256k1 private key. Handle as a secret. */
  privateKey: string;
  /** `0x`-prefixed 33-byte compressed secp256k1 public key. */
  publicKey: string;
  /** The ML-DSA-87 (Dilithium-5) keypair for the hybrid signature extension. */
  pqc: PqcKeypair;
}

/** Encode bytes to a lowercase `0x`-prefixed hex string. */
function toHexPrefixed(bytes: Uint8Array): string {
  let out = "0x";
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, "0");
  }
  return out;
}

/** Lowercase hex (no `0x`) of bytes. */
function toHex(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, "0");
  }
  return out;
}

/**
 * Render 20 address bytes as an EIP-55 mixed-case checksummed `0x` address.
 *
 * EIP-55: lowercase the hex address, take keccak256 of the lowercase hex (without
 * `0x`) as ASCII, and uppercase each hex nibble whose corresponding keccak nibble
 * is >= 8.
 */
function toEip55(addr20: Uint8Array): string {
  const lower = toHex(addr20);
  const hashHex = toHex(keccak_256(new TextEncoder().encode(lower)));
  let out = "0x";
  for (let i = 0; i < lower.length; i++) {
    out += parseInt(hashHex[i], 16) >= 8 ? lower[i].toUpperCase() : lower[i];
  }
  return out;
}

/**
 * Derive the three address encodings from a 20-byte account address.
 *
 * The SVM form is the base58 of the 32-byte address `addr20 ‖ 12 zero bytes`
 * (right-padded), matching the chain's unified 32-byte SVM address layout.
 *
 * @throws if `addr20` is not exactly 20 bytes.
 */
export function addressesFrom20(addr20: Uint8Array): UnifiedAddresses {
  if (addr20.length !== 20) {
    throw new Error(`address must be 20 bytes, got ${addr20.length}`);
  }
  const svmBytes = new Uint8Array(32);
  svmBytes.set(addr20, 0); // right-pad with 12 zero bytes → unified 32-byte SVM address
  return {
    addressBytes: addr20,
    cosmos: bytesToBech32(addr20, HRP),
    evm: toEip55(addr20),
    svm: base58.encode(svmBytes),
  };
}

/**
 * Decode any ONE of the three encodings of a unified account into all three.
 *
 * Provide exactly one of `cosmos` (`qor1…`), `evm` (`0x…`), or `hex` (20-byte
 * hex, with or without `0x`). The SVM base58 form is not accepted as input here
 * (it right-pads to 32 bytes; decode SVM externally to its 20-byte prefix if
 * needed).
 *
 * @throws if none of the inputs is provided, or the decoded payload is not 20 bytes.
 */
export function qoreAddresses(input: {
  cosmos?: string;
  evm?: string;
  hex?: string;
}): UnifiedAddresses {
  let addr20: Uint8Array;
  if (input.evm) {
    addr20 = hexToBytes(input.evm);
  } else if (input.hex) {
    addr20 = hexToBytes(input.hex);
  } else if (input.cosmos) {
    addr20 = bech32DataBytes(input.cosmos);
  } else {
    throw new Error("provide one of { cosmos, evm, hex }");
  }
  return addressesFrom20(addr20);
}

/** Parse a hex string (with or without `0x`) into bytes. */
function hexToBytes(hex: string): Uint8Array {
  const body = hex.startsWith("0x") || hex.startsWith("0X") ? hex.slice(2) : hex;
  if (body.length === 0 || body.length % 2 !== 0 || !/^[0-9a-fA-F]+$/.test(body)) {
    throw new Error(`invalid hex string: ${hex}`);
  }
  const bytes = new Uint8Array(body.length / 2);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(body.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/** Decode the raw byte payload of a bech32 (`qor1…`) address. `bech32ToHex`
 * returns `0x`-prefixed hex of the payload bytes. */
function bech32DataBytes(addr: string): Uint8Array {
  return hexToBytes(bech32ToHex(addr));
}

/**
 * The domain-separation context for the deterministic PQC seed. The 32-byte
 * SHAKE-256 of `"qorechain:pqc:v1|" + <cosmos-addr> + "|" + <secret>` is used as
 * the ML-DSA-87 keygen seed, so the PQC key is recoverable from the account and
 * the same secret. `<secret>` is the mnemonic for mnemonic-derived accounts and
 * the hex of the 32-byte seed for seed-derived accounts.
 */
const PQC_SEED_CONTEXT = "qorechain:pqc:v1|";

/** Derive the deterministic ML-DSA-87 keypair for an account. */
function derivePqc(cosmosAddr: string, secret: string): PqcKeypair {
  const seed = shake256(
    new TextEncoder().encode(`${PQC_SEED_CONTEXT}${cosmosAddr}|${secret}`),
    32,
  );
  return generatePqcKeypair(seed);
}

/** Build the address + PQC bundle from a 32-byte secp256k1 private key + secret. */
function accountFromPrivateKey(
  privkey: Uint8Array,
  pqcSecret: string,
): { addresses: UnifiedAddresses; publicKey: Uint8Array; pqc: PqcKeypair } {
  // Compressed (33B) public key for the SignerInfo pubkey; uncompressed (65B) for
  // the keccak address derivation.
  const compressed = secp256k1.getPublicKey(privkey, true); // 33 bytes
  const uncompressed = secp256k1.getPublicKey(privkey, false); // 65 bytes: 0x04||X||Y
  const addr20 = keccak_256(uncompressed.slice(1)).slice(12); // last 20 bytes
  const addresses = addressesFrom20(addr20);
  const pqc = derivePqc(addresses.cosmos, pqcSecret);
  return { addresses, publicKey: compressed, pqc };
}

/**
 * Derive a unified eth-native QoreChain account from a BIP-39 mnemonic.
 *
 * HD path: `m/44'/60'/0'/0/{index}` (SLIP-10 secp256k1). The 20-byte address is
 * `keccak256(uncompressedPubkey[1:])[12:]`; the PQC key is deterministically
 * derived from `shake256("qorechain:pqc:v1|" + cosmos + "|" + mnemonic, 32)`.
 *
 * @param mnemonic - A valid BIP-39 English mnemonic. Validated (checksum too).
 * @param index - Zero-based address index (last path segment). Defaults to `0`.
 * @throws if the mnemonic is invalid or `index` is not a non-negative integer.
 */
export async function deriveUnifiedAccount(
  mnemonic: string,
  index = 0,
): Promise<UnifiedAccount> {
  if (!validateMnemonic(mnemonic)) {
    throw new Error("invalid mnemonic");
  }
  if (!Number.isInteger(index) || index < 0) {
    throw new Error(`index must be a non-negative integer, got ${index}`);
  }
  const seed = await mnemonicToSeed(mnemonic);
  const node = HDKey.fromMasterSeed(seed).derive(`${ETH_HD_PATH_PREFIX}${index}`);
  if (!node.privateKey) {
    throw new Error("failed to derive secp256k1 key from mnemonic");
  }
  const privkey = node.privateKey;
  const { addresses, publicKey, pqc } = accountFromPrivateKey(privkey, mnemonic);
  return {
    mnemonic,
    privateKey: toHexPrefixed(privkey),
    publicKey: toHexPrefixed(publicKey),
    ...addresses,
    pqc,
  };
}

/**
 * Build a unified eth-native account directly from a 32-byte seed used AS the
 * secp256k1 private key (no HD derivation).
 *
 * Same address derivation as {@link deriveUnifiedAccount}; the PQC key is derived
 * from `shake256("qorechain:pqc:v1|" + cosmos + "|seed:" + hex(seed32), 32)` (note
 * the literal `"seed:"` prefix, so a seed-derived PQC key never collides with a
 * mnemonic-derived one). Use this for accounts anchored to a signature or an
 * externally supplied secret (see the Phantom P1a helper).
 *
 * @param seed32 - Exactly 32 bytes, used directly as the secp256k1 private key.
 * @throws if `seed32` is not 32 bytes or is not a valid secp256k1 scalar.
 */
export function unifiedAccountFromSeed(seed32: Uint8Array): UnifiedAccount {
  if (seed32.length !== 32) {
    throw new Error(`seed must be 32 bytes, got ${seed32.length}`);
  }
  if (!secp256k1.utils.isValidPrivateKey(seed32)) {
    throw new Error("seed is not a valid secp256k1 private key");
  }
  // Seed-derived accounts prefix the private-key hex with a literal "seed:" in
  // the PQC seed context (matches the reference wallet-adapter `walletFromSeed`),
  // so a seed-derived PQC key never collides with a mnemonic-derived one.
  const { addresses, publicKey, pqc } = accountFromPrivateKey(
    seed32,
    `seed:${toHex(seed32)}`,
  );
  return {
    privateKey: toHexPrefixed(seed32),
    publicKey: toHexPrefixed(publicKey),
    ...addresses,
    pqc,
  };
}

File: packages/rust/src/unified.rs
//! Unified eth-native QoreChain wallet.
//!
//! One eth-native secp256k1 keypair yields the SAME 20-byte account rendered as
//! all THREE QoreChain address encodings, so a wallet never "has funds on the
//! native lane but not the EVM lane" again. The 20 bytes are the Ethereum
//! derivation `keccak256(uncompressed_pubkey[1..])[12..]`, so the key is natively
//! spendable on the EVM lane; the native (`qor1…`) and SVM (base58) forms are
//! just other encodings of those same 20 bytes — the chain reads one balance for
//! the account, visible under all three.
//!
//! The account also carries a deterministic ML-DSA-87 (Dilithium-5) PQC keypair
//! for the on-chain hybrid ante, seeded from `{cosmos address, mnemonic}` so it
//! is recoverable without extra secret storage.
//!
//! This is ADDITIVE: the coin-118 native derivation in [`crate::accounts`] is
//! unchanged. Use this module when you want one identity shared across all lanes
//! and eth_secp256k1 signing ([`crate::sign_eth`]).

use crate::accounts::{derive_eth_key, NATIVE_PREFIX};
use crate::address::bytes_to_bech32;
use crate::error::Result;
use crate::pqc::{pqc_keypair_from_seed, PqcKeypair};
use k256::elliptic_curve::sec1::ToEncodedPoint;
use k256::SecretKey;
use sha3::digest::{ExtendableOutput, Update, XofReader};
use sha3::{Keccak256, Shake256};

/// The bech32 human-readable prefix for the unified account's native form.
pub const HRP: &str = NATIVE_PREFIX;

/// A unified eth-native QoreChain account.
///
/// The single secp256k1 keypair is rendered in all three address encodings, plus
/// a deterministic ML-DSA-87 keypair for the hybrid ante. Treat `private_key`
/// and `pqc.secret_key` as secrets.
#[derive(Debug, Clone)]
pub struct UnifiedAccount {
    /// 32-byte secp256k1 private key.
    pub private_key: [u8; 32],
    /// 33-byte compressed secp256k1 public key.
    pub public_key: Vec<u8>,
    /// The 20-byte account address (`keccak256(pubkey)[12..]`).
    pub address_bytes: [u8; 20],
    /// The native (bech32 `qor…`) encoding of the 20-byte address.
    pub cosmos: String,
    /// The EVM (EIP-55 `0x…`) encoding of the 20-byte address.
    pub evm: String,
    /// The SVM (base58 of `addr20 ‖ 12 zero bytes`) encoding.
    pub svm: String,
    /// The account's deterministic ML-DSA-87 (Dilithium-5) keypair.
    pub pqc: PqcKeypair,
}

/// The three address encodings of a 20-byte account (no key material).
#[derive(Debug, Clone)]
pub struct UnifiedAddresses {
    /// The 20-byte account address.
    pub address_bytes: [u8; 20],
    /// The native (bech32 `qor…`) form.
    pub cosmos: String,
    /// The EVM (EIP-55 `0x…`) form.
    pub evm: String,
    /// The SVM (base58 of `addr20 ‖ 12 zero bytes`) form.
    pub svm: String,
}

fn keccak256(data: &[u8]) -> [u8; 32] {
    use sha3::Digest;
    let out = Keccak256::digest(data);
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&out);
    arr
}

fn shake256_32(data: &[u8]) -> [u8; 32] {
    let mut hasher = Shake256::default();
    hasher.update(data);
    let mut reader = hasher.finalize_xof();
    let mut out = [0u8; 32];
    reader.read(&mut out);
    out
}

/// Renders 20 address bytes as a `0x`-prefixed EIP-55 mixed-case checksum address.
fn to_eip55(addr20: &[u8; 20]) -> String {
    let lower = hex::encode(addr20);
    let hash = keccak256(lower.as_bytes());
    let mut out = String::with_capacity(42);
    out.push_str("0x");
    for (i, c) in lower.chars().enumerate() {
        if c.is_ascii_digit() {
            out.push(c);
            continue;
        }
        let byte = hash[i / 2];
        let nibble = if i % 2 == 0 { byte >> 4 } else { byte & 0x0f };
        if nibble >= 8 {
            out.push(c.to_ascii_uppercase());
        } else {
            out.push(c);
        }
    }
    out
}

/// Derives the three encodings of a 20-byte account address.
///
/// - `cosmos` = bech32(`qor`, addr20)
/// - `evm`    = EIP-55(addr20)
/// - `svm`    = base58(addr20 ‖ 12 zero bytes) — the unified 32-byte SVM address
pub fn addresses_from_20(addr20: [u8; 20]) -> Result<UnifiedAddresses> {
    let cosmos = bytes_to_bech32(&addr20, HRP)?;
    let evm = to_eip55(&addr20);
    let mut svm_bytes = [0u8; 32];
    svm_bytes[..20].copy_from_slice(&addr20);
    let svm = bs58::encode(&svm_bytes).into_string();
    Ok(UnifiedAddresses {
        address_bytes: addr20,
        cosmos,
        evm,
        svm,
    })
}

/// Convenience: the three address encodings of an existing account, resolved
/// from any one of its `cosmos` (bech32), `evm`/`hex` (20-byte hex) forms.
///
/// Pass exactly one of the three; the first non-`None` argument wins in the
/// order `evm`, `hex`, `cosmos`.
pub fn qore_addresses(
    cosmos: Option<&str>,
    evm: Option<&str>,
    hex_addr: Option<&str>,
) -> Result<UnifiedAddresses> {
    let addr20 = if let Some(e) = evm {
        hex20_from_str(e)?
    } else if let Some(h) = hex_addr {
        hex20_from_str(h)?
    } else if let Some(c) = cosmos {
        let (_, data) =
            bech32::decode(c).map_err(|e| crate::error::Error::Address(e.to_string()))?;
        <[u8; 20]>::try_from(data.as_slice())
            .map_err(|_| crate::error::Error::Address("bech32 data must be 20 bytes".into()))?
    } else {
        return Err(crate::error::Error::Address(
            "provide one of {cosmos, evm, hex}".into(),
        ));
    };
    addresses_from_20(addr20)
}

fn hex20_from_str(s: &str) -> Result<[u8; 20]> {
    let trimmed = s.strip_prefix("0x").unwrap_or(s);
    let bytes = hex::decode(trimmed).map_err(|e| crate::error::Error::Address(e.to_string()))?;
    <[u8; 20]>::try_from(bytes.as_slice())
        .map_err(|_| crate::error::Error::Address("address must be 20 bytes".into()))
}

/// Derives the PQC seed for a unified account: `shake256("qorechain:pqc:v1|" +
/// cosmos + "|" + secret, 32)`, then seeds ML-DSA-87 keygen.
fn derive_pqc(cosmos: &str, secret: &str) -> PqcKeypair {
    let mut msg = Vec::new();
    msg.extend_from_slice(b"qorechain:pqc:v1|");
    msg.extend_from_slice(cosmos.as_bytes());
    msg.push(b'|');
    msg.extend_from_slice(secret.as_bytes());
    let seed = shake256_32(&msg);
    pqc_keypair_from_seed(&seed)
}

/// Builds a [`UnifiedAccount`] from a compressed/uncompressed secp256k1 pubkey,
/// a private key, and the `secret` string mixed into the PQC seed.
fn build_account(
    private_key: [u8; 32],
    compressed: Vec<u8>,
    uncompressed: &[u8],
    pqc_secret: &str,
) -> Result<UnifiedAccount> {
    // Ethereum address = last 20 bytes of keccak256 of the 64-byte pubkey body
    // (drop the 0x04 prefix).
    let digest = keccak256(&uncompressed[1..]);
    let mut addr20 = [0u8; 20];
    addr20.copy_from_slice(&digest[12..]);

    let enc = addresses_from_20(addr20)?;
    let pqc = derive_pqc(&enc.cosmos, pqc_secret);

    Ok(UnifiedAccount {
        private_key,
        public_key: compressed,
        address_bytes: addr20,
        cosmos: enc.cosmos,
        evm: enc.evm,
        svm: enc.svm,
        pqc,
    })
}

/// Derives the public keys (compressed 33B, uncompressed 65B) from a 32-byte
/// secp256k1 private key.
fn pubkeys_from_private(private_key: &[u8; 32]) -> Result<(Vec<u8>, Vec<u8>)> {
    let sk = SecretKey::from_slice(private_key)
        .map_err(|e| crate::error::Error::Derivation(format!("invalid private key: {e}")))?;
    let pk = sk.public_key();
    let compressed = pk.to_encoded_point(true).as_bytes().to_vec();
    let uncompressed = pk.to_encoded_point(false).as_bytes().to_vec();
    Ok((compressed, uncompressed))
}

/// Derives a unified eth-native account from a BIP-39 mnemonic at HD path
/// `m/44'/60'/0'/0/{index}`.
///
/// The PQC seed mixes the derived `cosmos` address with the `mnemonic`.
pub fn derive_unified_account(mnemonic: &str, index: u32) -> Result<UnifiedAccount> {
    let (private_key, compressed, uncompressed) = derive_eth_key(mnemonic, index)?;
    build_account(private_key, compressed, &uncompressed, mnemonic)
}

/// Builds a unified account directly from a 32-byte secp256k1 seed used as the
/// private key.
///
/// The PQC seed mixes the derived `cosmos` address with the `"seed:"`-prefixed
/// lowercase-hex encoding of the 32-byte private key (so a raw-seed account's PQC
/// key is still deterministic and recoverable from the seed alone).
///
/// # Safety of the seed
///
/// `seed32` **is** the account's spend key — everything else here is a rendering
/// of it. It MUST be real secret entropy: a CSPRNG draw, or a value derived from
/// one (a BIP-39 mnemonic via [`derive_unified_account`], an HSM, a KDF over a
/// secret the user alone holds).
///
/// It must NEVER be a wallet signature, or a hash of one, or any other value a
/// third party can ask a wallet, a service, or the user's browser to produce. A
/// wallet signature is a bearer secret: it is deterministic for a given message,
/// and any page that gets the wallet to sign that message receives the same bytes
/// — so whoever obtains them would hold this account's spend key. Link external
/// wallet keys through the authenticator lanes (`MsgRegisterAuthenticator` +
/// `MsgExecuteCosmos` / `MsgExecuteEVM`) instead, where an external signature
/// authorizes a spend but never becomes key material.
pub fn unified_account_from_seed(seed32: [u8; 32]) -> Result<UnifiedAccount> {
    let (compressed, uncompressed) = pubkeys_from_private(&seed32)?;
    build_account(
        seed32,
        compressed,
        &uncompressed,
        &format!("seed:{}", hex::encode(seed32)),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_MNEMONIC: &str =
        "test test test test test test test test test test test junk";

    #[test]
    fn kat_mnemonic_index0() {
        let acct = derive_unified_account(TEST_MNEMONIC, 0).unwrap();
        assert_eq!(acct.cosmos, "qor17w0adeg64ky0daxwd2ugyuneellmjgnxhkv37z");
        assert_eq!(acct.evm, "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266");
        assert_eq!(acct.svm, "HQ1S8pxTw4YN41GPdWYPQVfweQveXAUtfFnZNmvPfrYf");
        assert_eq!(acct.pqc.public_key.len(), 2592);
        assert_eq!(acct.pqc.secret_key.len(), 4896);
        // PQC public-key determinism (first 8 bytes) — locked against the 0.1.5
        // reference adapter to pin exact ML-DSA-87 keygen across languages.
        assert_eq!(hex::encode(&acct.pqc.public_key[..8]), "4a685622f2a99d54");
    }

    #[test]
    fn kat_seed_0x01() {
        let acct = unified_account_from_seed([0x01u8; 32]).unwrap();
        assert_eq!(acct.cosmos, "qor1rfjz7r3u8t65teavh5utquj3kwvsj983hh5zaj");
        assert_eq!(acct.evm, "0x1a642f0E3c3aF545E7AcBD38b07251B3990914F1");
        assert_eq!(acct.svm, "2n2Cnc7fib6rm4Azo1mbvMuHB3iCb1J254kEQDcrHyPu");
        assert_eq!(acct.pqc.public_key.len(), 2592);
        assert_eq!(acct.pqc.secret_key.len(), 4896);
        // PQC public-key determinism (first 8 bytes) — locked against the 0.1.5
        // reference adapter (uses the "seed:"-prefixed hex privkey in the PQC seed).
        assert_eq!(hex::encode(&acct.pqc.public_key[..8]), "2d7f888fecbe5b24");
    }

    #[test]
    fn addresses_from_20_roundtrip() {
        let acct = unified_account_from_seed([0x01u8; 32]).unwrap();
        let enc = addresses_from_20(acct.address_bytes).unwrap();
        assert_eq!(enc.cosmos, acct.cosmos);
        assert_eq!(enc.evm, acct.evm);
        assert_eq!(enc.svm, acct.svm);
    }

    #[test]
    fn qore_addresses_from_each_form() {
        let acct = unified_account_from_seed([0x01u8; 32]).unwrap();
        let from_evm = qore_addresses(None, Some(&acct.evm), None).unwrap();
        assert_eq!(from_evm.cosmos, acct.cosmos);
        let hex_addr = hex::encode(acct.address_bytes);
        let from_hex = qore_addresses(None, None, Some(&hex_addr)).unwrap();
        assert_eq!(from_hex.evm, acct.evm);
        let from_cosmos = qore_addresses(Some(&acct.cosmos), None, None).unwrap();
        assert_eq!(from_cosmos.svm, acct.svm);
    }

    #[test]
    fn pqc_deterministic() {
        let a = derive_unified_account(TEST_MNEMONIC, 0).unwrap();
        let b = derive_unified_account(TEST_MNEMONIC, 0).unwrap();
        assert_eq!(a.pqc.public_key, b.pqc.public_key);
        assert_eq!(a.pqc.secret_key, b.pqc.secret_key);
    }
}
